### DISCUSSION TITLE
Extract Derivatives from ADReal

### DISCUSSION THREAD

**Discussion by joe61vette** [idaholab/moose#23606] [Q&A General] [answered]
Hello:

I would like to get the pressure and temperature derivatives from the ADReal variable for density.  For example:

Real drho_dT = _rho[_qp].derivatives()[1];

This does give me the temperature derivative but only for the 1st quadrature point. So, the "1" appears to correspond to the index of the non-linear variable.  How can I get the indices of _T[_qp] and _p[_qp]?

Thanks,
  Joe Kelly

**Comment by GiudGiud** [comment, marked as answer]
Hello

There is an example in the INSFVMomentumAdvection fvkernel. First we get the dof number (here from the element argument)
    `const auto dof_number = sided_elem->dof_number(_sys.number(), _var.number(), 0);`
then access the derivatives at that index
    `const Real d_u_face_d_dof = u_face.derivatives()[dof_number];`

What kind of kernel or object are you using?
Are _T and _p both nonlinear variables?

Guillaume

**Reply by joe61vette** [reply]
Thanks Guillaume.  I knew there was some way to get the dof number but sometimes it is just hard to find.  Given that I would need to link to the variable number and then get the dof number, it is easier for me to just go ahead and add the eos functions that return the derivatives.  For example:

rho_from_p_T(Real p, Real T, Real & rho, Real & drho_dp, Real & drho_dT)

Thanks again,
  Joe

**Reply by GiudGiud** [reply]
This works.

In the context of fluid properties there's a few examples where the AD routine is used then the derivatives retrieved from that AD call manually. This is still not quite what you want but it could be a way to do `rho_from_p_T(Real p, Real T, Real & rho, Real & drho_dp, Real & drho_dT)` easily

For example:
```
void
Water97FluidProperties::T_from_p_h(
    Real pressure, Real enthalpy, Real & temperature, Real & dT_dp, Real & dT_dh) const
{
// this creates simple AD numbers with a 1 at a known index into the derivative array, for the dp_dp=1 here
  FPDualReal p = pressure;
  Moose::derivInsert(p.derivatives(), 0, 1.0);
  FPDualReal h = enthalpy;
  Moose::derivInsert(h.derivatives(), 1, 1.0);

// this is the AD routine. AD suffix is not needed if you implement an AD version of the routine manually
  const FPDualReal T = T_from_p_h_ad(p, h);

  temperature = T.value();
// we know the derivative indexing because we just set it above
  dT_dp = T.derivatives()[0];
  dT_dh = T.derivatives()[1];
}
```
